cmds/core/grep/grep.go

```diff
@@ -87,80 +91,125 @@ func grep(f *grepCommand, re *regexp.Regexp) {
 		lineNum++
 	}
 	close(res)
-	f.Close()
+	_ = f.rc.Close()
 }
 
-func printmatch(r *grepResult) {
+func (c *cmd) printMatch(r *grepResult) {
 	var prefix string
-	if r.match == !*invert {
-		matchCount++
+	if r.match == !c.params.invert {
+		c.matchCount++
 	}
-	if *count {
+	if c.params.count {
 		return
 	}
-	if showname {
-		fmt.Printf("%v", r.c.name)
+	if c.showName {
+		fmt.Fprintf(c.stdout, "%v", r.c.name)
 		prefix = ":"
 	}
-	if *noshowmatch {
-		fmt.Printf("\n")
+	if c.params.noShowMatch {
+		fmt.Fprintf(c.stdout, "\n")
 		return
 	}
-	if *number {
+	if c.params.number {
 		prefix = fmt.Sprintf(":%d:", r.lineNum)
 	}
-	if r.match == !*invert {
-		fmt.Printf("%v%v", prefix, *r.line)
+	if r.match == !c.params.invert {
+		fmt.Fprintf(c.stdout, "%v%v", prefix, *r.line)
+	}
+}
+
+type params struct {
+	expr            string
+	headers         bool
+	invert          bool
+	recursive       bool
+	noShowMatch     bool
+	count           bool
+	caseInsensitive bool
+	number          bool
+	quiet           bool
+}
+
+type cmd struct {
+	stdin      io.ReadCloser
+	stdout     io.Writer
+	stderr     io.Writer
+	allGrep    chan *oneGrep
+	params     params
```

Comment:
do you think we should embed the params struct? Then we could write c.invert, instead of c.params.invert, and so on. Not a big deal.

Reply:
yeah seems like good idea